apollo-router-core/src/query_planner/mod.rs

```diff
@@ -395,48 +356,108 @@ impl FetchNode {
-                }) => {
-                    let mut response = response
-                        .lock()
-                        .instrument(tracing::trace_span!("response_lock_wait"))
-                        .await;
+    async fn fetch_node<'a>(
+        &'a self,
+        data: &'a Value,
+        current_dir: &'a Path,
+        request: &'a Arc<Request>,
+        service_registry: Arc<dyn ServiceRegistry>,
+        schema: &'a Arc<Schema>,
+    ) -> Result<Value, FetchError> {
+        let FetchNode {
+            operation,
+            service_name,
+            ..
+        } = self;
 
-                    let span = tracing::trace_span!("response_insert");
-                    let _guard = span.enter();
-                    response.append_errors(&mut errors);
-                    response.insert_data(current_dir, data)?;
+        let query_span = tracing::info_span!("subfetch", service = service_name.as_str());
 
-                    Ok(())
-                }
-                None => Err(FetchError::SubrequestNoResponse {
+        let Variables { variables, paths } =
+            self.make_variables(data, current_dir, request, schema)?;
+
+        // We already checked that the service exists during planning
+        let fetcher = service_registry.get(service_name).unwrap();
+
+        let (res, _tail) = fetcher
+            .stream(
+                Request::builder()
+                    .query(operation)
+                    .variables(Some(Arc::new(variables)))
+                    .build(),
+            )
+            .await
+            .into_future()
+            .instrument(query_span)
+            .await;
+
+        let subgraph_response = match res {
+            Some(response) if !response.is_primary() => {
+                return Err(FetchError::SubrequestUnexpectedPatchResponse {
+                    service: service_name.to_owned(),
+                });
+            }
+            Some(subgraph_response) => subgraph_response,
+            None => {
+                return Err(FetchError::SubrequestNoResponse {
                     service: service_name.to_string(),
-                }),
+                })
             }
+        };
+
+        self.response_at_path(current_dir, paths, subgraph_response)
+    }
+
+    fn response_at_path<'a>(
+        &'a self,
+        current_dir: &'a Path,
+        paths: Option<Vec<Path>>,
+        subgraph_response: Response,
+    ) -> Result<Value, FetchError> {
+        let Response { data, .. } = subgraph_response;
+
+        if self.requires.is_some() {
+            // we have to nest conditions and do early returns here
+            // because we need to take ownership of the inner value
+            if let Value::Object(mut map) = data {
+                if let Some(entities) = map.remove("_entities") {
+                    tracing::trace!(
+                        "Received entities: {}",
+                        serde_json::to_string(&entities).unwrap(),
```

Comment:
:information_source:  This is evaluated even if the formatting is not done.

:screwdriver: Use `{:?}` instead. We don't need to format to json for debugging.

Reply:
that's potentially a big perf improvement :smiley:

Reply:
I didn't mean too :rofl:

Reply:
heh, accidentally faster is still faster, I'll take it :grin:

Reply:
done in 02b81b5